automerge-protocol/src/serde_impls/op.rs

```diff
@@ -53,6 +59,22 @@ impl Serialize for Op {
                 op.serialize_field("value", &value)?;
                 op.serialize_field("datatype", &DataType::Timestamp)?;
             }
+            OpType::Set(ScalarValue::Int(value)) => {
+                op.serialize_field("value", &value)?;
+                op.serialize_field("datatype", &DataType::Int)?;
+            }
+            OpType::Set(ScalarValue::Uint(value)) => {
+                op.serialize_field("value", &value)?;
+                op.serialize_field("datatype", &DataType::Uint)?;
+            }
+            OpType::Set(ScalarValue::F32(value)) => {
+                op.serialize_field("value", &value)?;
+                op.serialize_field("datatype", &DataType::F32)?;
+            }
+            OpType::Set(ScalarValue::F64(value)) => {
+                op.serialize_field("value", &value)?;
+                op.serialize_field("datatype", &DataType::F64)?;
+            }
             OpType::Set(value) => op.serialize_field("value", &value)?,
             OpType::MultiSet(values) => op.serialize_field("values", &values)?,
```

Comment:
Will this result in the `datatype` not being specified?

Reply:
Here's an example:

```json
{
  "ops": [
    {
      "action": "set",
      "obj": "_root",
      "elemId": "_head",
      "values": [
        "hello",
        3
      ],
      "pred": []
    }
  ],
  "actor": "",
  "seq": 0,
  "startOp": 0,
  "time": 0,
  "message": null,
  "deps": []
}
```

Reply:
I think the correct way to solve this is to add a serializer for `ScalarValue`s.

Reply:
How is this intended to work in general? Is the idea that all the `values` would be the same datatype?

Reply:
@alexjg I believe so

Reply:
@alexjg Is there any reason in particular why values in a `MultiSet` should have the same type? Otherwise I can remove some code I added here.

Reply:
Well if they are different types then we'll need to change the serialization logic to add a type tag to each value right? Otherwise how do you indicate that (for example) the first element in a list of numbers is a counter but the second is just an integer?

Reply:
Note that I am using the word `datatype` quite specifically to mean the `datatype` tag that is used for single element inserts

Reply:
I think MultiSet should focus on the case where all elements have the same datatype, since including a separate datatype tag for every value would add bloat, and the whole point of MultiSet is to be more compact. If the data contains values of different datatypes, they should be split into runs of MultiSet where each run only contains items of the same type (or use a plain Set if the run is of length 1).